lib/player.js

```diff
@@ -4371,6 +4467,102 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
     return shaka.util.StreamUtils.textStreamToTrack(stream);
   }
 
+  /**
+   * Adds the given chapters track to the loaded manifest.  <code>load()</code>
+   * must resolve before calling.  The presentation must have a duration.
+   *
+   * This returns the created track.
+   *
+   * @param {string} uri
+   * @param {string} language
+   * @param {string=} mimeType
+   * @return {!Promise.<shaka.extern.Track>}
+   * @export
+   */
+  async addChaptersTrack(uri, language, mimeType) {
+    if (this.loadMode_ != shaka.Player.LoadMode.MEDIA_SOURCE &&
+        this.loadMode_ != shaka.Player.LoadMode.SRC_EQUALS) {
+      shaka.log.error(
+          'Must call load() and wait for it to resolve before adding ' +
+          'chapters tracks.');
+      throw new shaka.util.Error(
+          shaka.util.Error.Severity.RECOVERABLE,
+          shaka.util.Error.Category.PLAYER,
+          shaka.util.Error.Code.CONTENT_NOT_LOADED);
+    }
+
+    if (!mimeType) {
+      // Try using the uri extension.
+      const extension = shaka.media.ManifestParser.getExtension(uri);
+      mimeType = shaka.Player.TEXT_EXTENSIONS_TO_MIME_TYPES_[extension];
+
+      if (!mimeType) {
+        try {
+          goog.asserts.assert(
+              this.networkingEngine_, 'Need networking engine.');
+          // eslint-disable-next-line require-atomic-updates
+          mimeType = await shaka.media.ManifestParser.getMimeType(uri,
+              this.networkingEngine_,
+              this.config_.streaming.retryParameters);
+        } catch (error) {}
+      }
+
+      if (!mimeType) {
+        shaka.log.error(
+            'The mimeType has not been provided and it could not be deduced ' +
+            'from its extension.');
+        throw new shaka.util.Error(
+            shaka.util.Error.Severity.RECOVERABLE,
+            shaka.util.Error.Category.TEXT,
+            shaka.util.Error.Code.TEXT_COULD_NOT_GUESS_MIME_TYPE,
+            extension);
+      }
+    }
+    if (mimeType != 'text/vtt') {
+      goog.asserts.assert(
+          this.networkingEngine_, 'Need networking engine.');
+      const data = await this.getTextData_(uri,
+          this.networkingEngine_,
+          this.config_.streaming.retryParameters);
+      const vvtText = this.convertToWebVTT_(data, mimeType);
+      const blob = new Blob([vvtText], {type: 'text/vtt'});
+      uri = shaka.media.MediaSourceEngine.createObjectURL(blob);
+      mimeType = 'text/vtt';
+    }
+    const trackElement = document.createElement('track');
+    trackElement.setAttribute('src', uri);
+    trackElement.setAttribute('kind', 'chapters');
+    trackElement.setAttribute('srclang', language);
+    // Because we're pulling in the text track file via Javascript, the
+    // same-origin policy applies. If you'd like to have a player served
```

Comment:
There is a lot of code duplication between this and addTextTrackAsync().  Can you please refactor?

Reply:
Done!